### PR TITLE
Show job template names if defined instead of test suites

### PR DESCRIPTION
**Pull request by Amrysliu**
This pull request has two purposes:

1. After we defined a test suite in YAML with an alias, the `Test result overview` page
and the `All Test` page will show the job template name (the new alias).

2. When triggering a job using `isos post`, the parameter `TEST` support using
job template name.

See: https://progress.opensuse.org/issues/59097

### CODE REVIEW

File: lib/OpenQA/Schema/Result/Jobs.pm [hunk 1]
```diff
@@ -305,7 +305,8 @@ sub label {
 sub scenario {
     my ($self) = @_;
 
-    return $self->result_source->schema->resultset('TestSuites')->find({name => $self->TEST});
+    my $test_suite_name = $self->settings_hash->{TEST_SUITE_NAME} || $self->TEST;
+    return $self->result_source->schema->resultset('TestSuites')->find({name => $test_suite_name});
```
Comment (okurz) [inline lib/OpenQA/Schema/Result/Jobs.pm]: I think this is a bit too UI-centric. Why not write the job template name into the TEST field when the job is *created*?

I think `$test_suite_name` can be misleading. There is no strong relation between TEST and the used test suite if any at all would have been used. E.g. if a job is just created with `openqa-client jobs post TEST=my_fancy_name` which does not relate to any test suite. Also there is the API to return a job and  there IMHO TEST should also reflect the final name. But as "name" is also already taken I do not have much ideas than `$test_suite_name`. Maybe `$template_name`?

File: lib/OpenQA/Schema/Result/ScheduledProducts.pm [hunk 2]
```diff
@@ -625,7 +629,6 @@ defined. Internal method used by the B<_schedule_iso()> method.
 
 sub _create_dependencies_for_job {
     my ($self, $job, $testsuite_mapping, $created_jobs, $cluster_parents, $skip_chained_deps) = @_;
-
```
Comment (okurz) [inline lib/OpenQA/Schema/Result/ScheduledProducts.pm]: unrelated change, please remove

File: lib/OpenQA/Schema/Result/ScheduledProducts.pm [hunk 3]
```diff
@@ -301,7 +301,6 @@ sub _schedule_iso {
                     });
             }
         }
-
```
Comment (okurz) [inline lib/OpenQA/Schema/Result/ScheduledProducts.pm]: unrelated change, please remove

File: lib/OpenQA/Schema/Result/ScheduledProducts.pm [hunk 4]
```diff
@@ -537,10 +536,11 @@ sub _generate_jobs {
             }
 
             # add properties from dedicated database columns to settings
-            $settings{TEST}              = $job_template->test_suite->name;
+            $settings{TEST}              = $job_template->name || $job_template->test_suite->name;
             $settings{MACHINE}           = $job_template->machine->name;
             $settings{BACKEND}           = $job_template->machine->backend;
             $settings{JOB_TEMPLATE_NAME} = $job_template->name if $job_template->name;
+            $settings{TEST_SUITE_NAME}   = $job_template->test_suite->name;
```
Comment (okurz) [inline lib/OpenQA/Schema/Result/ScheduledProducts.pm]: Can you explain why we need all three "TEST" and "JOB_TEMPLATE_NAME" and "TEST_SUITE_NAME"?

File: lib/OpenQA/Schema/Result/Jobs.pm (hunk 1 above)
Comment (Martchus) [reply lib/OpenQA/Schema/Result/Jobs.pm]: > I think this is a bit too UI-centric. Why not write the job template name into the TEST field when the job is created?
That's exactly what her code does and actually the reason why this modification in `sub scenario` is needed. Otherwise it wouldn't be possible to find the test suite anymore relying on `TEST`. Maybe that also explains your other question.

> I think $test_suite_name can be misleading. There is no strong relation between TEST and the used test suite if any at all would have been used.

But there is a relation which is now made more explicit (and the ` || $self->TEST` is only kept for backwards compatibility). And yes, this relation is only used to display the test suite description for the job so it is indeed not the "strongest" relation. However, I think the change makes still sense in order to keep things as functional as before.

File: lib/OpenQA/Schema/Result/Jobs.pm [hunk 5]
```diff
@@ -305,7 +305,8 @@ sub label {
 sub scenario {
     my ($self) = @_;
 
-    return $self->result_source->schema->resultset('TestSuites')->find({name => $self->TEST});
+    my $test_suite_name = $self->settings_hash->{TEST_SUITE_NAME} || $self->TEST;
```
[the remark this answers is not in the document]
Comment (okurz) [reply lib/OpenQA/Schema/Result/Jobs.pm]: yes, I do not worry about performance impact in this line but about our duplication and how it might be complicated and introduce bugs.

File: lib/OpenQA/Schema/Result/Jobs.pm (hunk 1 above)
Comment (okurz) [reply lib/OpenQA/Schema/Result/Jobs.pm]: @Martchus I don't think `$self->TEST` is only kept for backwards compatibility. I don't think we ever set TEST_SUITE_NAME e.g. in case of `openqa-client jobs post`, right?

Comment (Martchus) [reply lib/OpenQA/Schema/Result/Jobs.pm]: Right, but in these cases it is also questionable whether TEST corresponds to a test suite at all (so I didn't consider it). Note that so far we also don't set the JOB_TEMPLATE_NAME.

**Comment by Amrysliu** [comment]
> I suggest to add test cases for
> 
>     * single created jobs (e.g. "openqa-client jobs post") with a test suite name
> 
>     * display in test overview page, e.g. /tests/overview
> 
>     * display in /tests
> 
>     * display in /tests/[:id]

yes, I planed to add these test code after your review and be sure this way could make sense.

File: lib/OpenQA/Schema/Result/Jobs.pm (hunk 1 above)
Comment (Amrysliu) [reply lib/OpenQA/Schema/Result/Jobs.pm]: @okurz  `$self->TEST` is kept for backwards compatibility, and it's also used to deal with the situation that the job is created by `jobs post`. For `jobs post`, there is no `TEST_SUITE_NAME` and `JOB_TEMPLATE_NAME`, and in this context, the `TEST` is the test suite name.

@Martchus Thanks for your explanation.

File: lib/OpenQA/Schema/Result/Jobs.pm (hunk 5 above)
[the remark this answers is not in the document]
Comment (Amrysliu) [reply lib/OpenQA/Schema/Result/Jobs.pm]: @Martchus yes, you are right. I only find the `sub scenario` is called when showing one job's description.
@okurz I do not understand what the `our duplication` means, could you give me more detail?

File: lib/OpenQA/Schema/Result/ScheduledProducts.pm (hunk 4 above)
Comment (Amrysliu) [reply lib/OpenQA/Schema/Result/ScheduledProducts.pm]: For example, the YAML is defined like this:
```
scenarios:
  x86_64:
    sle-12-sp5-server-dvd-x86_64: 
    - apparmor_test:
        testsuite: apparmor
        settings:
          BETA: '1' 
    - apparmor_test2:
        testsuite: apparmor
        settings:
           BETA: '2'
```

File: lib/OpenQA/Schema/Result/Jobs.pm (hunk 5 above)
[the remark this answers is not in the document]
Comment (okurz) [reply lib/OpenQA/Schema/Result/Jobs.pm]: in the current version there is no real duplication anymore. We call "test_suite_name || test" here and later "job_template->name || test_suite->name" . I think it's ok.

[the remark this answers is not in the document]
Comment (okurz) [reply lib/OpenQA/Schema/Result/Jobs.pm]: oops, I found another occurence where you use "$settings{TEST} || $test eq $settings{TEST_SUITE_NAME}"

File: lib/OpenQA/Schema/Result/ScheduledProducts.pm (hunk 4 above)
Comment (okurz) [reply lib/OpenQA/Schema/Result/ScheduledProducts.pm]: Good that you mention it. For the second example of "START_AFTER_TEST" that should not be the test suite anymore but also the job template name. So for me this is another example why I am stating that the whole change is still too UI centric as with parameterized job template we should be changing the concept of how *every* job is created. Also "to find out" the test suite description I think we can try to read out a description from a job template and fallback to the test suite.

You are already on the right track but we have not completely followed through: In my opinion we should simply define the "TEST" based on the job template's name if it exists and rely on "TEST" everywhere else where we can. I know it's more than what we thought when you started but it just shows the general problems we still have not solved yet with job templates.

File: lib/OpenQA/Schema/Result/Jobs.pm (hunk 5 above)
[the remark this answers is not in the document]
Comment (Amrysliu) [reply lib/OpenQA/Schema/Result/Jobs.pm]: Do you mean the code in `sub _generate_jobs`?  this code is used to support to trigger job using `test suite name` and `job template name`. 
In the example I gave above, if I trigger job using `isos post TEST=apparmor`, there are two jobs will be created (variant1 and variant2 ). If I want to trigger one of them using `TEST`, our current code dose not support this operation. So I modify the code to support it. Users could trigger jobs use both test suite name and job template name.

File: lib/OpenQA/Schema/Result/ScheduledProducts.pm (hunk 4 above)
Comment (Amrysliu) [reply lib/OpenQA/Schema/Result/ScheduledProducts.pm]: > Good that you mention it. For the second example of "START_AFTER_TEST" that should not be the test suite anymore but also the job template name. So for me this is another example why I am stating that the whole change is still too UI centric as with parameterized job template we should be changing the concept of how _every_ job is created. Also "to find out" the test suite description I think we can try to read out a description from a job template and fallback to the test suite.

The `START_AFTER_TEST` is defined in test suite table. How could users know what the job template name when they add the test suite? So I do not think this value should be the job template name.

In my opinion, if we want to get the test suite description from a job template, it will be more complex.  If we change `TEST` to `JOB_TEMPLATE_NAME`, we do not know the real test suite name or id and how to find out the its description? 

> 
> You are already on the right track but we have not completely followed through: In my opinion we should simply define the "TEST" based on the job template's name if it exists and rely on "TEST" everywhere else where we can. I know it's more than what we thought when you started but it just shows the general problems we still have not solved yet with job templates.

Sorry, I do not understand your meaning clearly. In my code, I change the `TEST` to `JOB_TEMPLATE_NAME` when it exists, does it work? is there something I need to solve? If there is, feel free to tell me, I am very happy to do.

File: t/api/02-iso.t [hunk 6]
```diff
@@ -1297,4 +1301,73 @@ subtest '_SKIP_CHAINED_DEPS prevents scheduling parent tests' => sub {
...
+    my $schedule_param = {
+        ISO       => $iso,
+        DISTRI    => 'opensuse',
+        VERSION   => '13.1',
+        FLAVOR    => 'DVD',
+        ARCH      => 'i586',
+        BUILD     => '0091',
+        _GROUP_ID => '1002'
+    };
+    my $res = schedule_iso($schedule_param);
```
Comment (okurz) [inline t/api/02-iso.t]: I now just created https://github.com/os-autoinst/openQA/pull/2562 for this so we do not need to duplicate all the same default settings everywhere. If you agree we can merge my PR and then please use instead:

```perl
my $res = schedule_iso(%iso, _GROUP_ID => '1002');
```

of course you can still define `$schedule_param` as a copy to reuse it later or you write e.g.:

```perl
schedule_iso(%iso, TEST => 'child');
```

and such

File: t/api/02-iso.t [hunk 7]
```diff
@@ -1297,4 +1301,73 @@ subtest '_SKIP_CHAINED_DEPS prevents scheduling parent tests' => sub {
...
+    $res                                  = schedule_iso($schedule_param);
+    is($res->json->{count}, 2, 'do not schedule parent job');
+    my %schedule_jobs = map { $jobs->find($_)->TEST => 1 } @{$res->json->{ids}};
+    is_deeply(\%schedule_jobs, {child_variant1 => 1, child_variant2 => 1}, "only two children jobs were scheduled");
+
+    $schedule_param->{TEST} = 'child_variant1';
+    $res = schedule_iso($schedule_param);
+    is($res->json->{count}, 1, 'only schedule the child job with the same job template name');
+    my %child_job = map { $jobs->find($_)->TEST => 1 } @{$res->json->{ids}};
+    is_deeply(\%child_job, {child_variant1 => 1}, "only one child job were scheduled");
```
Comment (okurz) [inline t/api/02-iso.t]: s/were/was/

File: t/ui/01-list.t [hunk 8]
```diff
@@ -109,6 +109,27 @@ sub schema_hook {
...
+            t_created  => time2str('%Y-%m-%d %H:%M:%S', time - 3600, 'UTC'),
+            TEST       => 'kde_variant',
+            ARCH       => 'x86_64',
+            BUILD      => '0092',
+            DISTRI     => 'opensuse',
+            FLAVOR     => 'NET',
+            MACHINE    => '64bit',
+            VERSION    => '13.1',
+            settings =>
+              [{key => 'JOB_TEMPLATE_NAME', value => 'kde_variant'}, {key => 'TEST_SUITE_NAME', value => 'kde'}]});
```
Comment (okurz) [inline t/ui/01-list.t]: here I also recommend to reuse common settings with the job from L55 from a hash to not duplicate so much.

File: t/ui/01-list.t [hunk 9]
```diff
@@ -388,6 +409,18 @@ subtest 'check test results of job99940' => sub {
     }
 };
 
+subtest 'The TEST and description show correctly after changing TEST to JOB_TEMPLATE_NAME' => sub {
+    $driver->get('/tests');
+    my $td_99991 = $driver->find_element('#job_99991 td.test');
+    is($td_99991->get_text(), 'kde_variant@64bit', 'job 99991 displays TEST correctly');
```
Comment (okurz) [inline t/ui/01-list.t]: can you make that

```perl
is($driver->get_ok('/tests')->find_element('#job_99991 td.test')->get_text(), 'kde_variant@64bit', 'job 99991 displays TEST correctly');
```

File: t/ui/01-list.t [hunk 10]
```diff
@@ -388,6 +409,18 @@ subtest 'check test results of job99940' => sub {
     }
 };
 
+subtest 'The TEST and description show correctly after changing TEST to JOB_TEMPLATE_NAME' => sub {
```
Comment (okurz) [inline t/ui/01-list.t]: what do you mean with "changing"? Probably better something like

```perl
subtest 'test name and description still show up correctly using JOB_TEMPLATE_NAME' => sub {
```

File: t/ui/10-tests_overview.t [hunk 11]
```diff
@@ -319,6 +340,17 @@ subtest "filtering by machine" => sub {
...
+subtest "job template names displayed on 'Test result overview' page" => sub {
+    $driver->get('/group_overview/1002');
+    is($driver->find_element('.progress-bar-failed')->get_text(), '1 failed', 'The number of failed jobs is right');
+    is($driver->find_element('.progress-bar-unfinished')->get_text(),
+        '1 unfinished', 'The number of unfinished jobs is right');
+
+    $driver->get('/tests/overview?distri=opensuse&version=13.1&build=0091&groupid=1002');
+    my @tds = $driver->find_elements('#results_DVD tbody tr .name');
+    is($tds[0]->get_text(), 'kde_variant', 'job template name kde_variant displayed correctly');
+};
```
Comment (okurz) [inline t/ui/10-tests_overview.t]: I found out that on the job details page the test suite description is correctly shown for both kind but on the test overview page it is not shown for the parameterized job template.

File: t/api/02-iso.t (hunk 6 above)
Comment (Amrysliu) [reply t/api/02-iso.t]: After your PR has been merged, I will change this code.

File: t/ui/10-tests_overview.t (hunk 11 above)
Comment (Amrysliu) [reply t/ui/10-tests_overview.t]: Do you mean the description is not shown in `overview` page? our overview dose not show test suite description, right?  I do not see any test suite description is shown in the test over view page. For example `https://openqa.suse.de/tests/overview?distri=sle&version=12-SP5&build=0372&groupid=167`

File: lib/OpenQA/Schema/Result/ScheduledProducts.pm (hunk 4 above)
Comment (coolo) [reply lib/OpenQA/Schema/Result/ScheduledProducts.pm]: While what okurz writes is a little confusing, I agree:

It's called START_AFTER_TEST not START_AFTER_TESTSUITE_NAME - so if TEST comes from yaml, START_AFTER_TEST needs to follow.

Comment (Amrysliu) [reply lib/OpenQA/Schema/Result/ScheduledProducts.pm]: Sorry, I do not understand your meaning clearly.  If I understand correctly, whatever the `TEST` is test suite name or job template name, in Oliver's example, the job A won't be triggered.  I do not understand this sentence `The same we could apply by saying that we specify the final "job_template_name" in START_AFTER_TEST and it could be just the same a testsuite or a parameterized job template.`   Does it mean that we should use the `job_template_name` as the START_AFTER_TEST if the job_template_name exists?

File: t/api/02-iso.t (hunk 6 above)
Comment (okurz) [reply t/api/02-iso.t]: merged

File: t/ui/10-tests_overview.t (hunk 11 above)
Comment (okurz) [reply t/ui/10-tests_overview.t]: each test row that is corresponding to a test suite shows a link instead of just text. if you click on it you get a popover with the test suited description

File: lib/OpenQA/Schema/Result/ScheduledProducts.pm (hunk 4 above)
Comment (coolo) [reply lib/OpenQA/Schema/Result/ScheduledProducts.pm]: yes. And I think Oliver was giving a bad example - that's made it so confusing.

Comment (okurz) [reply lib/OpenQA/Schema/Result/ScheduledProducts.pm]: yes, this is what it means. Still I think we should try to update the value of TEST earlier, at the time of job creation, so that for START_AFTER_TEST we would not need to care where the value of TEST came from, if from a solely test suite based job or a parameterized job template one

Comment (Martchus) [reply lib/OpenQA/Schema/Result/ScheduledProducts.pm]: > so that for START_AFTER_TEST we would not need to care where the value of TEST came from

That would be ideal, indeed. And note that the same counts for the variables of the other dependency types.

File: lib/OpenQA/WebAPI/Controller/Test.pm [hunk 12]
```diff
@@ -467,12 +467,27 @@ sub prepare_job_results {
 
     # prefetch the number of available labels for those jobs
     my $job_labels = $self->_job_labels($jobs);
-    my @job_names  = map { $_->TEST } @$jobs;
+    my @job_names
+      = map { {job_template_name => $_->TEST, test_suite_name => $_->settings_hash->{TEST_SUITE_NAME} || $_->TEST} }
```
Comment (Martchus) [inline lib/OpenQA/WebAPI/Controller/Test.pm]: This might slow down the page loading time notably. See my suggestion in Rocket Chat.

Comment (Martchus) [reply lib/OpenQA/WebAPI/Controller/Test.pm]: I've just did a local test with OSD data. For a SLES12 build any only group "Functional: Server" the loading time increases from 0.9 seconds to 1.3 seconds. When showing the overview for the full build the loading time creases from 6.8 seconds to 11 seconds. So the performance impact is notable although it could be worse. There is likely still room for improvement.

Comment (Amrysliu) [reply lib/OpenQA/WebAPI/Controller/Test.pm]: @Martchus I have modified the code. Is this way your suggestion in Rocket Chat? I did the performance test, but I did not found the web page that spent 11 seconds. I did test using the url such as https://openqa.suse.de/tests/overview?distri=sle&version=12-SP5&build=0366&groupid=139

Comment (Amrysliu) [reply lib/OpenQA/WebAPI/Controller/Test.pm]: I did the test in my local and using OSD data. the performance is not good. It's almost equal to the last version.

Comment (Martchus) [reply lib/OpenQA/WebAPI/Controller/Test.pm]: The latest version is better in the regard that it doesn't query all job settings for a job. However, it still does individual queries per job and it might be possible to retrieve all settings for the relevant jobs in one query.

And now you're also querying the test suites individually for each job which looks counterproductive.

But I needed to play around with it myself to come up with better performing code.

Comment (Amrysliu) [reply lib/OpenQA/WebAPI/Controller/Test.pm]: @Martchus I modified the code to query the test suite one time. I did the performance test, it's better than the previous version. I do not mind that you take over the performance improvement, after all, I have spent a lot of time to do it and the performance is not good.  Thanks for your help.

**Comment by okurz** [comment]
After the impact of this PR is a bit bigger, could you please update the git commit message. The subject line starting with "Change" is probably not so helpful for a reader of the changelog. Think about what a user might want to read, e.g. you are adding a cool feature. Make sure people get it :)

File: lib/OpenQA/WebAPI/Controller/Test.pm (hunk 12 above)
Comment (Martchus) [reply lib/OpenQA/WebAPI/Controller/Test.pm]: I can play around with the query a little bit based on your current state.

**Comment by Martchus** [comment]
I rebased against the master and extracted parts of the loop into a separate function as it has grown a little bit big.

**Comment by Amrysliu** [comment]
> I pushed an additional commit to improve the performance. It is now only doing one additional query to the job settings table instead of one per job. Also note the use of map instead of loops.

It's much better than mine. Thanks for your improvement.

**Comment by okurz** [comment]
I think we only managed to come up with this nicer approach because Martchus and me were having the opportunity to sit together.
